internal/handlers/tigris/msg_find.go

```diff
@@ -16,11 +16,130 @@ package tigris
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // MsgFind implements HandlerInterface.
 func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
-	return nil, errNotImplemented
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	unimplementedFields := []string{
+		"skip",
+		"returnKey",
+		"showRecordId",
+		"tailable",
+		"oplogReplay",
+		"noCursorTimeout",
+		"awaitData",
+		"allowPartialResults",
+		"collation",
+		"allowDiskUse",
+		"let",
+	}
+	if err := common.Unimplemented(document, unimplementedFields...); err != nil {
+		return nil, err
+	}
+	ignoredFields := []string{
+		"hint",
+		"batchSize",
+		"singleBatch",
+		"maxTimeMS",
+		"readConcern",
+		"max",
+		"min",
+	}
+	common.Ignored(document, h.L, ignoredFields...)
+
+	var filter, sort, projection *types.Document
+	if filter, err = common.GetOptionalParam(document, "filter", filter); err != nil {
+		return nil, err
+	}
+	if sort, err = common.GetOptionalParam(document, "sort", sort); err != nil {
+		return nil, common.NewErrorMsg(common.ErrTypeMismatch, "Expected field sort to be of type object")
+	}
+	if projection, err = common.GetOptionalParam(document, "projection", projection); err != nil {
+		return nil, err
+	}
+
+	var limit int64
+	if l, _ := document.Get("limit"); l != nil {
+		if limit, err = common.GetWholeNumberParam(l); err != nil {
+			return nil, err
+		}
+	}
+
+	var fp fetchParam
+	if fp.db, err = common.GetRequiredParam[string](document, "$db"); err != nil {
+		return nil, err
+	}
+	collectionParam, err := document.Get(document.Command())
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	if fp.collection, ok = collectionParam.(string); !ok {
+		return nil, common.NewErrorMsg(
+			common.ErrBadValue,
+			fmt.Sprintf("collection name has invalid type %s", common.AliasFromType(collectionParam)),
+		)
+	}
+
+	fetchedDocs, err := h.fetch(ctx, fp)
+	if err != nil {
+		return nil, err
+	}
+
+	resDocs := make([]*types.Document, 0, 16)
+	for _, doc := range fetchedDocs {
+		matches, err := common.FilterDocument(doc, filter)
+		if err != nil {
+			return nil, err
+		}
+
+		if !matches {
+			continue
+		}
+
+		resDocs = append(resDocs, doc)
+	}
+
+	if err = common.SortDocuments(resDocs, sort); err != nil {
+		return nil, err
+	}
+	if resDocs, err = common.LimitDocuments(resDocs, limit); err != nil {
+		return nil, err
+	}
+	if err = common.ProjectDocuments(resDocs, projection); err != nil {
+		return nil, err
+	}
+
+	firstBatch := types.MakeArray(len(resDocs))
+	for _, doc := range resDocs {
+		if err = firstBatch.Append(doc); err != nil {
+			return nil, err
+		}
+	}
+
+	var reply wire.OpMsg
+	must.NoError(reply.SetSections(wire.OpMsgSection{
+		Documents: []*types.Document{must.NotFail(types.NewDocument(
+			"cursor", must.NotFail(types.NewDocument(
+				"firstBatch", firstBatch,
+				"id", int64(0), // TODO
+				"ns", fp.db+"."+fp.collection,
```

Comment:
I see `ns` mentioned here and there, and it's `db.collection`. I guess it means that neither db nor collection can have `.` as a part of their name?.. Is it validate anywhere (or I guess it might be a future task)?